src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -0,0 +1,122 @@
+        installed,  # type: Optional[Candidate]
+    ):
+        self._installed = installed
+        self._get_others = get_others
+        self._others = None  # type: Optional[Iterator[Candidate]]
+        self._returned = set()  # type: Set[_BaseVersion]
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._installed and self._installed.version not in self._returned:
+            self._returned.add(self._installed.version)
+            return self._installed
+        if self._others is None:
+            self._others = self._get_others()
+        cand = next(self._others)
+        while cand.version in self._returned:
+            cand = next(self._others)
+        self._returned.add(cand.version)
+        return cand
+
+    next = __next__  # XXX: Python 2.
+
+
+class _InstalledReplacesCandidatesIterator(collections_abc.Iterator):
+    """Iterator for ``FoundCandidates``.
+
+    This iterator is used when the resolver prefers to upgrade an
+    already-installed package. Candidates from index are returned in their
+    normal ordering, except replaced when the version is already installed.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+    ):
+        self._installed = installed
+        self._get_others = get_others
+        self._others = None  # type: Optional[Iterator[Candidate]]
+        self._returned = set()  # type: Set[_BaseVersion]
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._others is None:
+            self._others = self._get_others()
+        cand = next(self._others)
+        while cand.version in self._returned:
+            cand = next(self._others)
+        if self._installed and cand.version == self._installed.version:
+            cand = self._installed
+        self._returned.add(cand.version)
+        return cand
+
+    next = __next__  # XXX: Python 2.
+
+
+class FoundCandidates(collections_abc.Sequence):
+    """A lazy sequence to provide candidates to the resolver.
+
+    The intended usage is to return this from `find_matches()` so the resolver
+    can iterate through the sequence multiple times, but only access the index
+    page when remote packages are actually needed. This improve performances
+    when suitable candidates are already installed on disk.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+        prefers_installed,  # type: bool
+    ):
+        self._get_others = get_others
+        self._installed = installed
+        self._prefers_installed = prefers_installed
+
+    def __getitem__(self, index):
+        # type: (int) -> Candidate
+        for i, value in enumerate(self):
+            if index == i:
+                return value
+        raise IndexError(index)
```

Comment:
Maybe it's because I love micro-optimization, but this really bothers me.  I suggest going for something similar to the following:

```python
def __init__(self, ...) -> None:
    ...
    if prefers_installed:
        iterator = _InstalledFirstCandidatesIterator
    else:
        iterator = _InstalledReplacesCandidatesIterator
    self._iter = iterator(get_others, installed)
    self._candidates: List[Candidate] = []

@overload
def __getitem__(self, i: int) -> Candidate:
    if i < 0:
        self._candidates.extend(candidate)
    while i >= len(self._candidates):
        try:
            self._candidates.append(next(self._iter))
        except StopIteration:
            raise IndexError(index)  # from None
    return self._candidates[i]

@overload
def __getitem__(self, s: slice) -> Candidate:
    return [self[i] for i in range(s.start, s.stop, s.step)]

def __len__(self):
    self._candidates.extend(candidate)
    return len(self._candidates)
```

Reply:
The resolver does not actually use `__getitem__` anywhere, so I’d prefer not complicating things too much there. Maybe we should just put a `raise NotImplementedError()` in it to make things clear.

Caching resolved candidates is a good idea.

Reply:
I tried a couple of approaches to eagerly resolve the iterator class, but they all feel awkward to me since we still need `self._prefers_installed` for `__bool__`, and that means we have two attributes expressing the same thing that needs to be kept in sync. So I decided to keep it as is instead.

Reply:
Yea, I think keeping it as-it-is works well, and we can always iterate on this as we see fit going forward.